### ISSUE TITLE
My SqueezeNext implementation is not any better than my Tiny Darknet implementation

### ISSUE THREAD

**Issue by Fred-Erik** [amirgholami/SqueezeNext#3]
Hello!

Thank you all for the clear and complete paper you wrote! I managed to implement SqueezeNext in Tensorflow and am using it as a part of a CNN+RNN+Attention model to do OCR. Earlier, I designed the CNN part inspired by Tiny Darknet and made this structure:
```
type, filters, filter size, stride
conv 16 3x3 2x2
conv 32 3x3 2x2
conv 32 1x1 1x1
conv 64 3x3 1x1
conv 32 1x1 1x1
conv 128 3x3 2x1
conv 64 1x1 1x1
conv 256 3x3 1x1
conv 64 1x1 1x1
conv 512 3x3 2x1
conv 128 1x1 1x1
conv 1024 3x3 2x1
conv 64 1x1 2x1
```
Now, I implemented the 1.0-SqNxt-23v5 model (with some added strides to make sure the output is the same as the original model; from 128x64 input to 32x1 output). This results in about the same performance (in terms of accuracy and speed) on my target platform, an ARMv7 mobile chip (about the speed of a Raspberry Pi 2), and 1.5x slower on a GPU. Any intuitions if this result is to be expected? I would expect it to be either faster or give better results than my previous, simpler, model. Or should I just not expect your model to work as good on my very different dataset? (text lines)

Also, I noticed the training took a lot longer (about 3 times) to converge. Is that to be expected due to the massively increased network depth or is your network less optimal when running with batches on a GPU?

**Comment by amirgholami** [comment]
Hi Fred,

Thanks for reaching out. Without analyzing the particular dataset that you are using it is hard to give useful feedback. However, we have tested SqueezeNext on a range of tasks including style transfer and face detection. Note that for each application you will need to modify the neural network

**Comment by Fred-Erik** [comment]
Thank you so much for your response!

The data is IR and RGB cutouts of license plates. I sadly cannot give you the dataset because it is not legally allowed to share license plates due to privacy laws, but [here is an example image](https://imgur.com/rZ2fWXC) of my own license plate.

> However, we have tested SqueezeNext on a range of tasks including style transfer and face detection. Note that for each application you will need to modify the neural network
And what kind of modifications would be needed? I would suspect that because there is not one object like in ImageNet but multiple objects that have to be detected in both OCR and face detection, I would need more channels in the the beginning than in the ImageNet-optimized 1.0-SqNxt-23v5. Are those your findings as well, and/or would you be able to share the structure for the face detection network? 

> This results in about the same performance (in terms of accuracy and speed) on my target platform, an ARMv7 mobile chip
The performance in the paper is calculated on a simulated accelerator, have you by any chance also done tests on real-life low power hardware like an ARM processor?

**Comment by mrgloom** [comment]
I can be because Tiny Darknet just have less layers, it seems not only MACCs or FLOPS are matter but also memory access:

> The number of computations — whether you count them as MACCs or FLOPS — is only part of the story. Memory bandwidth is the other part, and most of the time is even more important!
> 
> On current computer architectures, a single memory access from main memory is much slower than a single computation — by a factor of about 100 or more!
> 
> You just saw that these neural networks do a lot of computations, but how many memory accesses do they perform?
> 
> For each layer, the device needs to:
> 
> read the input vector or feature map from main memory
> compute the dot products — which involves reading the layer’s weights from main memory too
> write the result back to main memory as a new vector or feature map.
> This involves a lot of memory accesses. Since memory is pretty slow, the amount of memory read/writes performed by the layer will have a big impact on its speed too — bigger perhaps than the number of computations.

https://machinethink.net/blog/how-fast-is-my-model/

btw `CNN+RNN+Attention model to do OCR` any reference to this kind of models?
